File: services/instagram-scraper/worker.py
"""
RabbitMQ worker for consuming crawl requests using pika.

This worker:
- Consumes from 'crawl_requests' queue
- Processes Instagram URLs using InstagramCrawler
- Publishes results to 'raw_recipe_data' queue
"""
import pika
import json
import logging
import signal
import sys
from instagram_crawler import InstagramCrawler
from models import CrawlRequest, RawRecipeData
from config import RABBITMQ_HOST

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class CrawlWorker:
    """Worker that processes Instagram crawl requests from RabbitMQ"""

    def __init__(self):
        self.connection = None
        self.channel = None
        self.crawler = InstagramCrawler()
        self.should_stop = False

    def connect(self):
        """Establish connection to RabbitMQ"""
        logger.info(f"Connecting to RabbitMQ at {RABBITMQ_HOST}")
        self.connection = pika.BlockingConnection(
            pika.ConnectionParameters(host=RABBITMQ_HOST)
        )
        self.channel = self.connection.channel()

        # Declare queues (durable for persistence)
        self.channel.queue_declare(queue='crawl_requests', durable=True)
        self.channel.queue_declare(queue='raw_recipe_data', durable=True)

        # Process one message at a time (QoS)
        self.channel.basic_qos(prefetch_count=1)

        logger.info("Connected to RabbitMQ successfully")

    def publish_raw_recipe_data(self, raw_data: RawRecipeData):
        """Publish extracted data to raw_recipe_data queue"""
        try:
            logger.info(f"Publishing raw recipe data for post: {raw_data.url}")

            # Prepare message
            message = raw_data.model_dump()
            message['timestamp'] = raw_data.timestamp.isoformat()

            # Publish with persistence
            self.channel.basic_publish(
                exchange='',
                routing_key='raw_recipe_data',
                body=json.dumps(message),
                properties=pika.BasicProperties(
                    delivery_mode=2,  # Make message persistent
                    content_type='application/json'
                )
            )

            logger.info(f"Successfully published recipe data: {raw_data.url}")

        except Exception as exc:
            logger.error(f"Failed to publish raw recipe data: {exc}")
            raise

    def process_message(self, ch, method, properties, body):
        """Process a single crawl request message"""
        try:
            # Parse message
            raw_data = json.loads(body.decode())
            logger.debug(f"Received message: {raw_data} (type: {type(raw_data)})")

            # Handle both dict and list (in case of old message format)
            if isinstance(raw_data, list):
                # If it's a list, take the first element
                if len(raw_data) > 0 and isinstance(raw_data[0], dict):
                    request_data = raw_data[0]
                    logger.warning("Received message as list, using first element")
                else:
                    raise ValueError(f"Invalid message format: expected dict or list of dicts, got {type(raw_data)}")
            elif isinstance(raw_data, dict):
                request_data = raw_data
            else:
                raise ValueError(f"Invalid message format: expected dict or list, got {type(raw_data)}")

            logger.info(f"Processing crawl request: {request_data.get('instagram_url', 'unknown')}")

            # Validate request data
            request = CrawlRequest(**request_data)

            # Extract post data
            raw_data = self.crawler.extract_post_data(str(request.instagram_url))

            # Publish to next queue
            self.publish_raw_recipe_data(raw_data)

            # Acknowledge message after successful processing
            ch.basic_ack(delivery_tag=method.delivery_tag)

            logger.info(f"Successfully processed: {request.instagram_url}")

        except Exception as exc:
            logger.error(f"Failed to process message: {exc}", exc_info=True)

            # Reject and requeue for retry (could implement max retry logic here)
            ch.basic_nack(delivery_tag=method.delivery_tag, requeue=True)

    def start(self):
        """Start consuming messages"""
        logger.info("Starting worker...")

        # Set up signal handlers for graceful shutdown
        signal.signal(signal.SIGINT, self.handle_shutdown)
        signal.signal(signal.SIGTERM, self.handle_shutdown)

        # Connect to RabbitMQ
        self.connect()

        # Start consuming
        logger.info("Waiting for messages from 'crawl_requests' queue. Press CTRL+C to exit.")
        self.channel.basic_consume(
            queue='crawl_requests',
            on_message_callback=self.process_message,
            auto_ack=False  # Manual acknowledgment for reliability
        )

        try:
            self.channel.start_consuming()
        except KeyboardInterrupt:
            self.stop()

    def handle_shutdown(self, signum, frame):
        """Handle shutdown signals"""
        logger.info(f"Received signal {signum}, shutting down gracefully...")
        self.stop()

    def stop(self):
        """Stop consuming and close connections"""
        if self.channel and self.channel.is_open:
            logger.info("Stopping consumer...")
            self.channel.stop_consuming()

        if self.connection and self.connection.is_open:
            logger.info("Closing connection...")
            self.connection.close()

        logger.info("Worker stopped")
        sys.exit(0)


def main():
    """Entry point for the worker"""
    worker = CrawlWorker()
    worker.start()


if __name__ == '__main__':
    main()

File: services/recipe-schema-converter/src/worker.ts
/**
 * RabbitMQ worker for consuming raw recipe data and publishing structured recipes.
 */

import * as amqp from "amqplib";
import { config } from "./config.ts";
import { RecipeConverter } from "./converter.ts";
import type { RawRecipeData } from "./types/index.ts";

export class RecipeConverterWorker {
  private connection: amqp.Connection | null = null;
  private channel: amqp.Channel | null = null;
  private converter: RecipeConverter;

  constructor() {
    this.converter = new RecipeConverter();
  }

  /**
   * Connect to RabbitMQ and set up channels/queues.
   */
  async connect(): Promise<void> {
    console.log(`Connecting to RabbitMQ at ${config.rabbitmq.host}...`);

    this.connection = await amqp.connect(config.rabbitmq.url);
    this.channel = await this.connection.createChannel();

    // Declare queues
    await this.channel.assertQueue(config.queues.input, { durable: true });
    await this.channel.assertQueue(config.queues.output, { durable: true });
    await this.channel.assertQueue(config.queues.failed, { durable: true });

    // Set prefetch for fair dispatch
    await this.channel.prefetch(config.worker.prefetchCount);

    console.log("Connected to RabbitMQ successfully");
  }

  /**
   * Start consuming messages.
   */
  async start(): Promise<void> {
    console.log("Starting Recipe Schema Converter Worker...");

    // Setup graceful shutdown
    process.on("SIGINT", () => this.stop());
    process.on("SIGTERM", () => this.stop());

    await this.connect();

    console.log(`Waiting for messages from '${config.queues.input}' queue...`);

    await this.channel!.consume(
      config.queues.input,
      (msg) => this.processMessage(msg),
      { noAck: false }
    );
  }

  /**
   * Process a single message.
   */
  private async processMessage(msg: amqp.ConsumeMessage | null): Promise<void> {
    if (!msg || !this.channel) return;

    let rawData: RawRecipeData | null = null;

    try {
      rawData = JSON.parse(msg.content.toString()) as RawRecipeData;
      console.log(`Processing recipe from: ${rawData.url}`);

      // Convert using LLM
      const structuredRecipe = await this.converter.convert(rawData);

      // Publish to output queue
      this.channel.sendToQueue(
        config.queues.output,
        Buffer.from(JSON.stringify(structuredRecipe)),
        { persistent: true, contentType: "application/json" }
      );

      // Acknowledge successful processing
      this.channel.ack(msg);
      console.log(
        `Successfully converted: ${rawData.url} (confidence: ${structuredRecipe.extraction_confidence})`
      );
    } catch (error) {
      console.error("Error processing message:", error);

      const retryCount = this.getRetryCount(msg);

      if (retryCount < config.worker.maxRetries) {
        await this.scheduleRetry(msg, retryCount);
      } else {
        await this.moveToFailedQueue(msg, rawData, error);
      }

      // Acknowledge to remove from original queue
      this.channel.ack(msg);
    }
  }

  /**
   * Get the current retry count from message headers.
   */
  private getRetryCount(msg: amqp.ConsumeMessage): number {
    const headers = msg.properties.headers || {};
    return (headers["x-retry-count"] as number) || 0;
  }

  /**
   * Schedule a retry with exponential backoff.
   */
  private async scheduleRetry(
    msg: amqp.ConsumeMessage,
    retryCount: number
  ): Promise<void> {
    const delay =
      config.worker.retryDelays[
        Math.min(retryCount, config.worker.retryDelays.length - 1)
      ];

    console.log(
      `Scheduling retry ${retryCount + 1}/${config.worker.maxRetries} in ${delay / 1000}s`
    );

    // Simple retry: wait and republish to input queue
    // For production, consider using a delayed message exchange plugin
    setTimeout(() => {
      this.channel?.sendToQueue(config.queues.input, msg.content, {
        persistent: true,
        contentType: "application/json",
        headers: {
          ...msg.properties.headers,
          "x-retry-count": retryCount + 1,
        },
      });
    }, delay);
  }

  /**
   * Move failed message to dead letter queue.
   */
  private async moveToFailedQueue(
    msg: amqp.ConsumeMessage,
    rawData: RawRecipeData | null,
    error: unknown
  ): Promise<void> {
    const failedMessage = {
      original_message: rawData || JSON.parse(msg.content.toString()),
      error: error instanceof Error ? error.message : String(error),
      stack: error instanceof Error ? error.stack : undefined,
      retry_count: this.getRetryCount(msg),
      failed_at: new Date().toISOString(),
    };

    this.channel?.sendToQueue(
      config.queues.failed,
      Buffer.from(JSON.stringify(failedMessage)),
      { persistent: true, contentType: "application/json" }
    );

    console.log(`Message moved to failed queue after ${config.worker.maxRetries} retries`);
  }

  /**
   * Gracefully stop the worker.
   */
  async stop(): Promise<void> {
    console.log("Shutting down worker...");

    if (this.channel) {
      await this.channel.close();
    }
    if (this.connection) {
      await this.connection.close();
    }

    console.log("Worker stopped");
    process.exit(0);
  }
}